### PR TITLE
[alt] adds all onnx node tests

### PR DESCRIPTION
**Pull request by zjgarvey**
to run one of these tests, you could try:

```python
python run.py --cachedir=${CAHCE_DIR} -t test_av -v
```

to see the results for avg pool, for example.

### CODE REVIEW

File: alt_e2eshark/onnx_tests/operators/generate_node.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class NodeTest(OnnxModelInfo):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.model = onnx_node_tests_dir + self.name + "/model.onnx"
+
+    def construct_inputs(self):
+        inputpb = onnx_node_tests_dir + self.name + "/test_data_set_0/input_0.pb"
+        return get_tensor_from_pb(inputpb)
```
[the remark this answers is not in the document]
Comment (zjgarvey) [reply alt_e2eshark/onnx_tests/operators/generate_node.py:35]: Yeah, I just noticed this when trying to run all of them. Updating now

**Comment by ScottTodd** [comment]
While there isn't much code here (yet), I'd prefer for us to lean on the existing setup at https://github.com/nod-ai/SHARK-TestSuite/tree/main/iree_tests/onnx for these particular tests. iree_tests already has robust support for xfails and setting up test configurations (cpu / gpu / lists of flags).

File: alt_e2eshark/onnx_tests/operators/generate_node.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+def get_tensor_from_pb(inputpb):
+    proto = TensorProto()
+    with open(inputpb, "rb") as f:
+        proto.ParseFromString(f.read())
+    t = numpy_helper.to_array(proto)
+    return TestTensors((t,))
+
+
+onnx_node_tests_dir = getsitepackages()[0] + "/onnx/backend/test/data/node/"
```
[the remark this answers is not in the document]
Comment (zjgarvey) [reply alt_e2eshark/onnx_tests/operators/generate_node.py:23]: Oh, nice!

**Comment by zjgarvey** [comment]
> While there isn't much code here (yet), I'd prefer for us to lean on the existing setup at https://github.com/nod-ai/SHARK-TestSuite/tree/main/iree_tests/onnx for these particular tests. iree_tests already has robust support for xfails and setting up test configurations (cpu / gpu / lists of flags).

Yeah, I agree (especially for CI jobs or running many tests). I thought it would still be helpful to add these node tests in case someone (e.g. a torch-mlir developer without a local iree-build) wants to debug an iree_tests failure using this test suite setup. I ended up writing this after working with Alex to test his TFIDF changes in torch-mlir and it was really helpful. 

Let me add a few small changes to make this work out better (e.g for multiple inputs), and I'll put in a comment to mention the iree_tests/onnx as an alternative.

**Comment by ScottTodd** [comment]
Yep, makes sense for working with torch-mlir. This might make sense to contribute upstream to torch-mlir itself in some way for that case. @renxida was looking into that too.

I don't want to end up with multiple ways to run the same test suite, each with their own quirks. I also have some ideas for how to bring python test sources in to iree_tests but I want to be really careful there. The core workflows there need to be as easy as possible to integrate with upstream tools (`iree-compile`, `iree-run-module`), ideally across platforms (including where a python interpreter can't run).

File: alt_e2eshark/onnx_tests/operators/generate_node.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zjgarvey) [reply alt_e2eshark/onnx_tests/operators/generate_node.py]: I'm just going to leave it as a comment for now (how to use the git submodule instead of the python installation).

File: alt_e2eshark/onnx_tests/operators/generate_node.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zjgarvey) [reply alt_e2eshark/onnx_tests/operators/generate_node.py]: I don't think I'm going to add extra tests for the other test_data_set's yet.

**Comment by zjgarvey** [comment]
> Yep, makes sense for working with torch-mlir. This might make sense to contribute upstream to torch-mlir itself in some way for that case. @renxida was looking into that too.
> 
> I don't want to end up with multiple ways to run the same test suite, each with their own quirks. I also have some ideas for how to bring python test sources in to iree_tests but I want to be really careful there. The core workflows there need to be as easy as possible to integrate with upstream tools (`iree-compile`, `iree-run-module`), ideally across platforms (including where a python interpreter can't run).

So do you think these tests shouldn't be added? I don't really see an issue with adding them (this is all of the code required for doing so), and it would be helpful for torch-mlir devs who don't want to build iree locally when debugging iree_test failures.

**Comment by zjgarvey** [comment]
> > and it would be helpful for torch-mlir devs who don't want to build iree locally when debugging iree_test failures.
> 
> The more I think about this the more I think these tests should just go in torch-mlir itself (https://github.com/llvm/torch-mlir or a new repo in the llvm org), not in nod-ai/SHARK-TestSuite. I want to move iree_tests to iree-org/SOME_REPO_NAME too. Contributors shouldn't need to switch github orgs to run tests for their code.

That's a good point. It might make the most sense to add the onnx frontend stuff into torch-mlir's e2e testing and just run it as part of their testing setup. Until that happens (it would be a bit of a project), should I leave this as a draft PR? Add it to the repo but comment out the test registration and just let it be a demo? This work is still useful in the meantime, so I'd prefer not to just throw it away.

**Comment by renxida** [comment]
Major difference with https://github.com/llvm/torch-mlir/tree/main/projects/pt1/python/torch_mlir_e2e_test/test_suite 

is that those tests go from torch->onnx->our onnx.operator "dialect" ->torch-mlir

and has numerous limitations including:
- some onnx ops are simply untestable because we are unable to express them in torch
- some onnx optional attributes / inputs are untestable because similar reasons
- intermediate outputs are difficult to produce and replication is super annoying
- MOST IMPORTANTLY, many onnx layers can't get numerically tested because the project/pt1 test suite re-randomizes the weights and biases between the reference run and the torch-mlir run!!!

I'm pretty sure Zach's stuff solves all of the above problems and more, and I feel strongly that we should supplant the onnx testing in torch-mlir with his work.

**Comment by zjgarvey** [comment]
> I'd stage the torch-mlir + ONNX tests as a standalone test suite without over-generalizing, to help put it on a path to being upstreamed.

Yeah, that would make sense. We can use torch-mlir's default backend instead of iree, and it completely removes that dependency from testing torch-mlir conversions. 

> 
> Actually, how does this differ from these?
> 
>     * https://github.com/llvm/torch-mlir/blob/main/projects/pt1/python/torch_mlir_e2e_test/configs/onnx_backend.py
> 
>     * https://github.com/llvm/torch-mlir/blob/main/projects/pt1/e2e_testing/xfail_sets.py
> 
>     * https://github.com/llvm/torch-mlir/tree/main/projects/pt1/python/torch_mlir_e2e_test/test_suite
> 
> 
> Looks like those tests are all defined there, while these use the upstream onnx test cases.

The e2e tests in torch mlir all start from a `torch.nn.Module`. the onnx config in that test runner will use `torch.onnx.export` to generate an onnx model, which it then imports to torch-mlir. It compares the torch-mlir/llvm result against pytorch's native inference result. Many of these tests fail because pytorch's onnx exporter doesn't support converting a particular pytorch op to onnx. Torch-mlir devs working on the ONNX frontend are really only concerned about getting a correct result starting from an onnx model/ onnx op, so the current roundabout strategy is a bit annoying. 

In contrast, the starting point for the tests here is an onnx model (either downloaded/found/generated with helpers), and the gold reference here is generated with onnxruntime. 

> 
> To put this on a path to be usable by regular torch-mlir contributors, it should match the existing test structure in that repo as closely as possible. Docs like https://github.com/llvm/torch-mlir/blob/main/docs/adding_an_e2e_test.md should then also be updated.

I actually tried modeling some of the structure in this test-suite after the testing infrastructure in torch-mlir, so I imagine it is possible to integrate this there. @renxida mentioned some interest in figuring out how to set this up and we could certainly work together to get it working. 

----------------------

This conversation brings me to a broader discussion: 

### "What is the point of e2eshark?" 
1. We *want* to test framework(specifically ONNX)->MLIR->IREE workflows. It is likely that IREE will not be a testing dependency for torch-mlir, so the idea of adding this type of a testing infrastructure to torch-mlir isn't completely sufficient as a replacement.
2. We want to start with an onnx model, testing both large models and individual ops. This is not set-up in torch-mlir yet, and will take some time to add.
3. We want to dump a bunch of artifacts and reproduction information.
4. Manage the downloading of models from Azure, so that testing azure models is as easy as possible. 

### "What is the point of iree_tests, and how does it's role differ from e2eshark?"
My understanding is that iree_tests start with an mlir file together with inputs/gold ouputs. This way it can be tested completely from within IREE, and requires likely 0 external dependencies. This sounds like a very efficient design for CI jobs, and is particularly useful for iree developers. 

It sounds like you have a script for generating/updating the node tests automatically, too. Therefore, adding the node tests here isn't *necessary*.

### "Why do I want to add these node tests to e2eshark?"
1. There is no cost here. Redundancy of these node tests with the iree_tests/onnx is not a cost unless this test suite is going to be used as a CI job somewhere. In my opinion, even after some significant improvements to this test-suite, it would still only make sense to use this to test the many public/private azure models, because iree_tests will always be a more efficient alternative for testing operators (and probably several models, too). 
2. There is currently no good replacement for this in iree_tests, from the perspective of the torch-mlir developers in the nod team. Until a replacement for the node tests is added to torch-mlir like we discussed above, it still makes sense to just add these here. The onnx frontend torch-mlir folks are pushing on getting better e2e results right now, and I think that this patch would be helpful for them immediately. 

### "Why do you NOT want to add these node tests here?"
To try and summarize what I THINK your position is:

"They are redundant with iree_tests. The advantage provided by adding them here would more appropriately belong in torch-mlir's native testing infrastructure".

I also assume that adding the node tests here would reduce the urgency for updating the torch-mlir testing infrastructure, so adding this could delay the up-streaming of something useful to the broader torch-mlir community. 

### What should we do right now?
I certainly agree that the e2e testing in torch-mlir needs to be fleshed out, and I don't want to cause delay in that effort, but I also don't want to delay progress for our developers in the meantime. Therefore, it makes the most sense to me to:
1.  submit an issue in torch-mlir with some basic outlining of an e2e testing infrastructure change.
2.  merge this PR so our nod developers can use it right now if they want. 
3.  follow up on updating the torch-mlir testing infrastructure with @renxida .

-----------------

Sorry for the long comment, @ScottTodd . Am I correct in my characterization of `iree_tests` and of your objection to merging this pr?

**Comment by ScottTodd** [comment]
I'm in agreement with most of that last message.

Here are my general concerns, some of which aren't specific to this PR:

* I don't want us to build an island of test suites that only nod-ai / AMD employees can contribute to. If work is happening in upstream repositories somewhere, we should find ways to bring the tests as close as possible to the code. See the README in SHARK-TestSuite: "Temporary home of a test suite we are evaluating". It's getting to be time to find a graduation path for the tests.
  * Our group that nod-ai / AMD has interests in running tests using private models. That work should probably be happening in private repositories or off of GitHub entirely. I'm not comfortable mixing public and private data, logs, code, etc. from either a security or a developer experience perspective.
  * Mixing public and private code tests also makes upstreaming harder
* Certain types of tests are too large or complicated to put directly in a project repository. Some of that is just keeping things organized (1000s of loose files can be noisy in git fetching / history, requiring all sorts of dependencies), and some is technical limitations of Git / Git LFS (megabytes/gigabytes of data).

---

> 1. There is no cost here. Redundancy of these node tests with the iree_tests/onnx is not a cost unless this test suite is going to be used as a CI job somewhere.

Agreed. I like the ONNX node tests in particular since they are fully defined upstream and we're just consuming them for our downstream testing. Having multiple implementations of test suites using the same raw files is totally fine. What I _do_ want to avoid is duplicate tracking of pass/fails through similar or slightly different workflows. We've already had a _lot_ of duplicated work done in https://github.com/nod-ai/SHARK-TestSuite/tree/main/e2eshark/onnx/operators - most of those are likely very similar to https://github.com/onnx/onnx/tree/main/onnx/backend/test/case/node.

---

Your next step plans sound good to me, but let's make sure we do the work to get code upstream and deprecate/delete anything we're hoping to replace. This sort of code tends to rot if left unmaintained.

**Review by ScottTodd — APPROVED** [review]
